librz/util/log.c

```diff
@@ -5,133 +5,207 @@
-static int cfg_logtraplvl = RZ_LOGLVL_FATAL; // Log trap level
-static bool cfg_logsrcinfo = false; // Print out debug source info with the output
-static bool cfg_logcolors = false; // Output colored log text based on level
-static char cfg_logfile[LOG_CONFIGSTR_SIZE] = ""; // Output text to filename
-static const char *level_tags[] = { // Log level to tag string lookup array
-	[RZ_LOGLVL_SILLY] = "SILLY",
+typedef struct log_config_s {
+	RzList *callbacks;
+	RzLogLevel level;
+	RzLogLevel trap_level;
+	bool show_sources;
+	FILE *file;
+	const char **tags;
+	RzThreadLock *lock;
+} log_config_t;
+
+///< Log level to tag string lookup array
+static const char *level_tags_no_colors[] = {
 	[RZ_LOGLVL_VERBOSE] = "VERBOSE",
 	[RZ_LOGLVL_DEBUG] = "DEBUG",
 	[RZ_LOGLVL_INFO] = "INFO",
 	[RZ_LOGLVL_WARN] = "WARNING",
 	[RZ_LOGLVL_ERROR] = "ERROR",
-	[RZ_LOGLVL_FATAL] = "FATAL"
+	[RZ_LOGLVL_FATAL] = "FATAL",
 };
 
+static const char *level_tags_colors[] = {
+	[RZ_LOGLVL_VERBOSE] = Color_GREEN "VERBOSE" Color_RESET,
+	[RZ_LOGLVL_DEBUG] = Color_BLUE "DEBUG" Color_RESET,
+	[RZ_LOGLVL_INFO] = Color_CYAN "INFO" Color_RESET,
+	[RZ_LOGLVL_WARN] = Color_YELLOW "WARNING" Color_RESET,
+	[RZ_LOGLVL_ERROR] = Color_RED "ERROR" Color_RESET,
+	[RZ_LOGLVL_FATAL] = Color_MAGENTA "FATAL" Color_RESET,
+};
+
+static log_config_t logcfg = { 0 };
+
+static void log_init() {
+	if (logcfg.lock) {
+		return;
+	}
+	logcfg.callbacks = NULL;
+	logcfg.level = RZ_DEFAULT_LOGLVL;
+	logcfg.trap_level = RZ_DEFAULT_LOGLVL_TRAP;
+	logcfg.show_sources = false;
+	logcfg.file = NULL;
+	logcfg.tags = level_tags_no_colors;
+	logcfg.lock = rz_th_lock_new(false);
+}
+
 // cconfig.c configuration callback functions below
 RZ_API void rz_log_set_level(RzLogLevel level) {
-	cfg_loglvl = level;
+	log_init();
+	rz_th_lock_enter(logcfg.lock);
+	logcfg.level = level;
+	rz_th_lock_leave(logcfg.lock);
 }
 
 RZ_API void rz_log_set_traplevel(RzLogLevel level) {
-	cfg_logtraplvl = level;
+	log_init();
+	rz_th_lock_enter(logcfg.lock);
+	logcfg.trap_level = level;
+	rz_th_lock_leave(logcfg.lock);
 }
 
-RZ_API void rz_log_set_file(const char *filename) {
-	int value_len = rz_str_nlen(filename, LOG_CONFIGSTR_SIZE) + 1;
-	strncpy(cfg_logfile, filename, value_len);
+RZ_API bool rz_log_set_file(RZ_NULLABLE const char *filename) {
+	log_init();
+	rz_th_lock_enter(logcfg.lock);
+	bool ret = true;
+	if (logcfg.file) {
+		// if already open, then close the file handler.
+		fclose(logcfg.file);
+		logcfg.file = NULL;
+	}
+
+	if (RZ_STR_ISEMPTY(filename)) {
+		// allow to have no filename
+		goto end;
+	}
+
+	FILE *file = rz_sys_fopen(filename, "a+");
+	if (!file) {
+		file = rz_sys_fopen(filename, "w+");
+	}
+	if (!file) {
+		// failed to open the file.
+		ret = false;
+		goto end;
+	}
+	logcfg.file = file;
+	ret = true;
+
+end:
+	rz_th_lock_leave(logcfg.lock);
+	return ret;
 }
 
-RZ_API void rz_log_set_srcinfo(bool show_info) {
-	cfg_logsrcinfo = show_info;
+RZ_API void rz_log_set_show_sources(bool show_sources) {
```

Comment:
Could you please add Doxygen to the API functions that miss it while you are on it?